Test lib fixes

IgorMinar opened the pull request.

IgorMinar committed:
style(di): fix a variable name typo

IgorMinar committed:
refactor(collection.ts): simplify ListWrapper.clear implementation

IgorMinar committed:
fix(fake_async): remove unused variable

IgorMinar committed:
fix(test_lib): reexport fake_async via angular/test

previously fake_async was not being publically exported at all

IgorMinar committed:
fix(build): add config for outputting the missing test_lib.d.ts file

jeffbcross commented:
Looks good once the issues are fixed.

Here's the branch commit of the work we started, @IgorMinar https://github.com/jeffbcross/angular/commit/076507154e2ea6c737a67393f52075ba51f59a54

jeffbcross commented:
BTW I removed a `comp:` label because @mhevery's dashboard assumes only one component label.

IgorMinar committed:
fix(dts generation): add support for type aliases

IgorMinar committed:
fix(test_lib): add missing types

IgorMinar committed:
fix(dts generation): rewrite the d.ts file code generator to fix bugs and apply type remap correctly

Previously the type remap was not being applied to comments and free floating functions.

The nunjucks template was becoming unreadable so rather than making a tweak there I
rewrote it into imperative code that is much easier to follow.

The output was diffed against the old output. The diff contained only the expected changes.

IgorMinar commented:
there seems to be a flaky test that intermittently fails on travis. AFAIK it is not being introduced by my changes, so I'll send this to the presubmit queue as is.

mary-poppins commented:
Merging PR #3952 on behalf of @IgorMinar to branch [presubmit-IgorMinar-pr-3952](https://github.com/angular/angular/tree/presubmit-IgorMinar-pr-3952).
